src/context.rs

```diff
@@ -65,6 +72,24 @@ impl<'ctx> MessageContext<'ctx> {
         self.terms.get(id)
     }
 
+    pub fn add_function<F>(&mut self, id: &str, func: F)
+    where
+        F: 'ctx + Fn(&[Option<FluentValue>], &HashMap<String, FluentValue>) -> Option<FluentValue>,
+    {
+        self.functions.insert(id.to_string(), Box::new(func));
+    }
```

Comment:
In JS we pass functions into the constructor of the `MessageContext`. Is there any reason you moved it to a separate function? I'd prefer to stay consistent with JS when possible.

Reply:
Actually, I'd argue that `add_function` is a good design for Rust. It's consistent with how you populate `HashMaps` via `insert()`. And it reduces the boilerplate. The alternative would be to pass a `HashMap` of functions into the constructor. In order to build the `HashMap` you'd need to `insert()` every function separately anyways. `add_function` abstracts that away.

Reply:
oh, ok! Convinced!

Reply:
I agree with @stasm. I tried making the changes that you suggested yesterday @zbraniecki but I encountered the same issues that @stasm brought up.